File: unity/Assets/Scripts/unityagent/Operation/EgressSystemMessageSender.cs
using System.Net.Sockets;
using System.Threading.Tasks;

using UnityAgent.Buffers;
using UnityAgent.Management.Message;
using UnityAgent.SystemMessage;

namespace UnityAgent.Operation
{
public class EgressSystemMessageSender : WorkerThread
{
    private NetworkStream stream;

    private class Nested
    {
        static Nested()
        {
        }

        internal static readonly EgressSystemMessageSender instance = new();
    }

    private EgressSystemMessageSender() : base("ESM Sender")
    {
    }

    public static EgressSystemMessageSender Instance => Nested.instance;

    public bool HasSentGoOfflineMessage()
    {
        return HasStopped();
    }

    public bool HasStopped()
    {
        return task.Status == TaskStatus.Canceled ||
               task.Status == TaskStatus.Faulted ||
               task.Status == TaskStatus.RanToCompletion;
    }

    public void Run(NetworkStream stream)
    {
        this.stream = stream;
        task = Task.Run(() => Send());
    }

    private void Send()
    {
        try
        {
            while (!CancellationToken.IsCancellationRequested)
                SendEgressSystemMessage();
            SendGoOfflineMessageToAllNetworkNodes();
        }
        catch
        {
            HandleConnectionClosedException();
        }
        CancellationToken.ThrowIfCancellationRequested();
    }

    private void SendEgressSystemMessage()
    {
        EgressSystemMessage esm = DequeueEsm();
        if (esm.Metadata != null)
        {
            SystemMessageMetadata m = esm.Metadata;
            try
            {
                RoutingInfo ri = esm.RoutingInfo;
                RoutingManagementData rmd = ri.RoutingManagementData;
            }
            catch
            {
            }
        }
        byte[] data = esm.Serialize();
        stream.Write(data, 0, data.Length);
    }

    private EgressSystemMessage DequeueEsm()
    {
        EgressSystemMessageBuffer esmBuffer = EgressBuffer.Dequeue();
        if (IsHeartbeat(esmBuffer))
            return HeartbeatMessage.Message;
        return esmBuffer.Dequeue();
    }

    private bool IsHeartbeat(EgressSystemMessageBuffer buffer)
    {
        return buffer == null;
    }

    private void SendGoOfflineMessageToAllNetworkNodes()
    {
        EgressSystemMessage esm = CreateGoOfflineMessage();
        byte[] data = esm.Serialize();
        stream.Write(data, 0, data.Length);
    }

    private EgressSystemMessage CreateGoOfflineMessage()
    {
        RoutingData routingData = new();
        RoutingManagementData rmd = new GoOffline(Agent.Instance.AgentId);
        RoutingInfo routingInfo = new(routingData, rmd);
        EgressSystemMessage esm = new(routingInfo);
        return esm;
    }
}
}

File: ros/ros_packages/rosagent/rosagent/operation/egress_system_message_sender.py
from threading import Thread

from rosagent.buffers import egress_buffer as EgressBuffer
from rosagent.message.heartbeat_message import HeartbeatMessage

from rosutility.socket.socket_transceiver import SocketTransceiver

class EgressSystemMessageSender(Thread):

    _instance = None

    def __new__(cls, socket):
        if cls._instance is None:
            cls._instance = super(EgressSystemMessageSender, cls).__new__(cls)
            cls._instance._initialized = False
        return cls._instance

    def __init__(self, socket):
        if self._initialized: return
        self._initialized = True
        Thread.__init__(self)
        self.daemon = True
        self.socket = socket
        self._socket_transceiver = SocketTransceiver(socket)
        self._heartbeat_message = HeartbeatMessage()

    def run(self):
        while True:
            self.send_esm()

    def send_esm(self):
        esm = self.dequeue_esm()
        if esm is None: return
        esm_bytes = esm.serialize()
        self._socket_transceiver.send(esm_bytes)
    
    def dequeue_esm(self):
        esm_buffer = EgressBuffer.dequeue()
        if self.is_heartbeat(esm_buffer):
            esm = self._heartbeat_message.get()
        else:
            esm = esm_buffer.dequeue()
        return esm
    
    def is_heartbeat(self, value):
        return value is None
